grove/plugin/range_req_handler.go

```diff
@@ -0,0 +1,225 @@
+			sep = "&"
+		}
+		newKey := d.DefaultCacheKey + sep + "grove_range_req_handler_plugin_data=" + d.Req.Header.Get("Range")
+		overRideFunc(newKey)
+		log.Debugf("range_req_handler: store_ranges default key:%s, new key:%s", d.DefaultCacheKey, newKey)
+	}
+}
+
+// rangeReqHandleBeforeParent changes the parent request if needed (mode == get_full_serve_range)
+func rangeReqHandleBeforeParent(icfg interface{}, d BeforeParentRequestData) {
+	log.Debugf("rangeReqHandleBeforeParent calling.")
+	rHeader := d.Req.Header.Get("Range")
+	if rHeader == "" {
+		log.Debugf("No Range header found")
+		return
+	}
+	log.Debugf("Range string is: %s", rHeader)
+	cfg, ok := icfg.(*rangeRequestConfig)
+	if !ok {
+		log.Errorf("range_req_handler config '%v' type '%T' expected *rangeRequestConfig\n", icfg, icfg)
+		return
+	}
+	if cfg.Mode == "get_full_serve_range" {
+		// get_full_serve_range means get the whole thing from parent/org, but serve the requested range. Just remove the Range header from the upstream request
+		d.Req.Header.Del("Range")
+	}
+	return
+}
+
+// rangeReqHandleBeforeRespond builds the 206 response
+// Assume all the needed ranges have been put in cache before, which is the truth for "get_full_serve_range" mode which gets the whole object into cache.
+func rangeReqHandleBeforeRespond(icfg interface{}, d BeforeRespondData) {
+	log.Debugf("rangeReqHandleBeforeRespond calling\n")
+	ictx := d.Context
+	ctx, ok := (*ictx).([]byteRange)
+	if !ok {
+		log.Errorf("Invalid context: %v", ictx)
+	}
+	if len(ctx) == 0 {
+		return // there was no (valid) range header
+	}
+
+	cfg, ok := icfg.(*rangeRequestConfig)
+	if !ok {
+		log.Errorf("range_req_handler config '%v' type '%T' expected *rangeRequestConfig\n", icfg, icfg)
+		return
+	}
+	if cfg.Mode == "store_ranges" {
+		return // no need to do anything here.
+	}
+
+	multipartBoundaryString := ""
+	originalContentType := d.Hdr.Get("Content-type")
+	*d.Hdr = web.CopyHeader(*d.Hdr) // copy the headers, we don't want to mod the cacheObj
+	if len(ctx) > 1 {
+		//multipart = true
+		multipartBoundaryBytes := make([]byte, 8)
+		if _, err := rand.Read(multipartBoundaryBytes); err != nil {
+			log.Errorf("Error with rand.Read: %v", err)
+		}
+		multipartBoundaryString = hex.EncodeToString(multipartBoundaryBytes)
+		d.Hdr.Set("Content-Type", fmt.Sprintf("multipart/byteranges; boundary=%s", multipartBoundaryString))
+	}
+	totalContentLength, err := strconv.ParseInt(d.Hdr.Get("Content-Length"), 10, 64)
+	if err != nil {
+		log.Errorf("Invalid Content-Length header: %v", d.Hdr.Get("Content-Length"))
+	}
+	body := make([]byte, 0)
+	for _, thisRange := range ctx {
+		if thisRange.End == -1 || thisRange.End >= totalContentLength { // if the end range is "", or too large serve until the end
+			thisRange.End = totalContentLength - 1
+		}
+		log.Debugf("range:%d-%d", thisRange.Start, thisRange.End)
+		if multipartBoundaryString != "" {
+			body = append(body, []byte(fmt.Sprintf("\r\n--%s\r\n", multipartBoundaryString))...)
```

Comment:
Golang Printf functions are incredibly slow. I know the code is uglier, but I'd recommend using string concatenation and `strconv.Itoa`, possibly even consider `bytes.Buffer`. This is in the request path, so it'll likely directly impact the number of requests Grove can handle concurrently.